### ISSUE TITLE
Not sure whether backend attributes are accessible

### ISSUE THREAD

**Issue by ysb33r** [asciidoctor/asciidoctorj#316]
NOTE: This is working against the `asciidoctorj-1.6.0` branch. Given that I was experiementing with a converter in Groovy, the sample code is in Groovy below.

Given that I have the following converter

```groovy
class MyConverter extends AbstractConverter {

    MyConverter(final String backend,Map<Object, Object> opts) {
        super(backend, opts)
    }

    Object convert(AbstractNode node, String transform, Map<Object, Object> opts) {
        if (node instanceof DocumentRuby) {
            println node.document.options // Prints OK
            println node.document.attributes // Prints OK
            println node.document.options.attributes.docdir // Fails with null object
            return null
        } else {
            return null
        }
    }
}
```

This is an example of what `node.document.options` will look like
```
[  backend:leanpub, 
   header_footer:true, 
   to_dir:/Users/schalkc/.../test/leanpub, 
   attributes:[docfile:/Users/schalkc/.../test/test-documents/book-with-prematter.adoc, docdir:/Users/schalkc/.../test/test-documents, docname:book-with-prematter, docdate:2015-05-25, doctime:19:23:14 BST, docdatetime:2015-05-25 19:23:14 BST]
]
```

There you can see the existence of `attributes`. However when trying to access it, it will fail with attributes being `null`. It will also report the type of attributes to be `java.lang.Boolean`.

I suspect that the attributes entry is not translated correctly from Ruby to Java.

**Comment by mojavelinux** [comment]
We don't carry over the options into the document object, so therefore the option on Document is not the same as the options passed to the API. This is an Asciidoctor core issue.

There's some awkwardness around this API...since it wasn't really designed as part of the API. A related discussion is here: https://github.com/asciidoctor/asciidoctor/issues/692

**Comment by mojavelinux** [comment]
In short, the path `node.document.options.attributes` is not part of the public API.

**Comment by mojavelinux** [comment]
What attribute are you trying to retrieve?

**Comment by ysb33r** [comment]
`docdir`. (As `document.attributes.docdir` is empty).

I needed that to work out a default images directory in case `imagesdir` was not set or was relative. (Code I roughly translated form Ruby in the way the epub3 converter works out the location of the frontcover image).

**Comment by mojavelinux** [comment]
docdir is available as `document.attr 'docdir'` if the safe mode is < SERVER. See https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/document.rb#L307-L350

**Comment by mojavelinux** [comment]
...have said that, you can also access it via document.base_dir (though there are some differences).

**Comment by mojavelinux** [comment]
> (Code I roughly translated form Ruby in the way the epub3 converter works out the location of the frontcover image).

This logic is actually incorrect. If you are reading the image (and not linking to it), you can find the correct logic in the PDF converter.

https://github.com/asciidoctor/asciidoctor-pdf/blob/master/lib/asciidoctor-pdf/converter.rb#L1748-L1792

Locating images is still too much of a burden for extension writers. But we have to see more use cases to understand how to make good APIs for this. Suggestions welcome of course.

**Comment by mojavelinux** [comment]
I want to point out that there is a big difference in how images are resolved if you are linking to them vs if you are embedding them. This is one of the things that makes this API challenging to get right.

**Comment by ysb33r** [comment]
> docdir is available as document.attr 'docdir' if the safe mode is < SERVER. See https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/document.rb#L307-L350

It still turns out to be empty.

> ...have said that, you can also access it via document.base_dir (though there are some differences).

`base_dir` does not seem to be in the Java API.

**Comment by mojavelinux** [comment]
> It still turns out to be empty.

That's strange. Is the key set and the value empty or is the key missing?

> base_dir does not seem to be in the Java API.

That should be fixed then.

**Comment by ysb33r** [comment]
The key is there

```
[encoding:UTF-8, sectids:, toc-placement:auto, stylesheet:, webfonts:, prewrap:, attribute-undefined:drop-line, attribute-missing:skip, iconfont-remote:, caution-caption:Caution, important-caption:Important, note-caption:Note, tip-caption:Tip, warning-caption:Warning, appendix-caption:Appendix, example-caption:Example, figure-caption:Figure, table-caption:Table, toc-title:Table of Contents, manname-title:NAME, untitled-label:Untitled, version-label:Version, last-update-label:Last updated, docfile:book-with-prematter.adoc, 
  docdir:, 
  docname:book-with-prematter, docdate:2015-05-25, doctime:22:24:52 BST, docdatetime:2015-05-25 22:24:52 BST, asciidoctor:, asciidoctor-version:1.5.2, safe-mode-name:secure, safe-mode-secure:, safe-mode-level:20, max-include-depth:64, user-home:., backend:leanpub, linkcss:, doctype:article, doctype-article:, backend-leanpub-doctype-article:, backend-leanpub:, outfilesuffix:.html, filetype:html, filetype-html:, basebackend:leanpub, basebackend-leanpub:, basebackend-leanpub-doctype-article:, localdate:2015-05-25, localtime:22:44:22 BST, localdatetime:2015-05-25 22:44:22 BST, stylesdir:., iconsdir:./images/icons, doctitle:Title, firstname:Asciidoctor, author:Asciidoctor Team, authorinitials:AT, lastname:Team, email:foo@bar.example, authorcount:1, authors:Asciidoctor Team, leanpub_lang:afrikaans, imagesdir:./build/resources/test/test-documents/images, front-cover-image:image:Simple.png[]]
```
As some background Asciidoctor was instantiated in this way

https://github.com/ysb33r/asciidoctor-leanpub-converter/blob/master/core/src/test/groovy/org/ysb33r/asciidoctor/leanpub/internal/LeanpubSpecification.groovy

**Comment by mojavelinux** [comment]
Are you sure it is blank even when you set the `safe` option to `safe` (or `unsafe`)?

https://github.com/ysb33r/asciidoctor-leanpub-converter/blob/master/core/src/test/groovy/org/ysb33r/asciidoctor/leanpub/internal/LeanpubSpecification.groovy#L54

This scenario should be a test in AsciidoctorJ (very that docdir is set if safe is set...so we know safe is being set properly).

**Comment by ysb33r** [comment]
Adding 'safe : 0 // or 1 or 2', causes `docdir` to be populated.  Thus leaving it undefined is causing `docdir` to be null. 

In `options.rb` of `asciidoctor`, the default value will be `SafeMode::UNSAFE` if not set. I suspect that similar behaviour is not replicated in asciidoctorj.

**Comment by mojavelinux** [comment]
When using Asciidoctor via the API, the default safe mode is SECURE. This is a hard requirement for Asciidoctor to be used on GitHub. docdir must also be blank. AsciidoctorJ respects these defaults.

When using Asciidoctor via the CLI or plugins, the default safe mode is unsafe.

I know this is a bit confusing, but that's why we always recommend being explicit about the safe mode when using the API.

**Comment by mojavelinux** [comment]
Btw, string values are permitted (unsafe, safe, server, secure).

**Comment by robertpanzer** [comment]
I tried to reproduce it.

This works for me:
`System.out.println(((Map) node.getOptions().get("attributes")).get("docdir"));`

But this fails with the same NPE:
`System.out.println(((Map) node.getDocument().getOptions().get("attributes")).get("docdir"));`

node is already a Document, why do you call getDocument()?

But there are some points where we still have problems and don't convert all the way down.
For example node.getDocument() will return a Ruby object instead of the DocumentRuby instance.
This is something that we will have to resolve.
Do you think it is necessary that node.getDocument() always returns the *same* object as the DocumentRuby object passed originally to convert()?
I hope the answer is No :)

**Comment by mojavelinux** [comment]
> For example node.getDocument() will return a Ruby object instead of the DocumentRuby instance.
This is something that we will have to resolve.

Thanks for getting to the heart of the issue here, Robert! Now it makes sense.

From AsciidoctorJ, getDocument() should return the AsciidoctorJ document. You often have to use this in abstracts when you aren't sure if the node is the root (the document) or not. If you're already on the Document, then it just returns self / this (speaking in pseudo-code).

**Comment by mojavelinux** [comment]
I do see from the Ruby API that options on Document is a public instance variable. However, it's really intended to be for internal use only because it doesn't give a complete picture. That's something we'll need to look into...probably as part of the issue I linked to above.

**Comment by robertpanzer** [comment]
This will work as well:
`((Map) document.getDocument().getOptions().get(JRubyRuntimeContext.get().newSymbol("attributes"))).get("docdir")`

The problem is apparently that getOptions() returns a RubyHash on the Ruby document.
This can only be addressed with RubySymbols :-(

**Comment by robertpanzer** [comment]
I could possibly even return the same instances for any node.getDocument() if I cache the converted objects in instance vars of the Ruby nodes.
That means that I could set `Asciidoctor::AbstractNode#asciidoctorJCache` to a `RubyHash` where I link to the converter AsciidoctorJ Node.

@mojavelinux What do you think about this? Is it worth to try?

**Comment by ysb33r** [comment]
> node is already a Document, why do you call getDocument()?

I did not know any better at the time.

**Comment by robertpanzer** [comment]
I am not very fluent with this API myself yet, so maybe there was a good reason to do so. :)

**Comment by ysb33r** [comment]
Maybe the following is related. Given this document,

```
= Working with Images

[chapter]
== Block image

=== Block image with role

.Image One Left
[.left.text-center]
image::Image1.png[]
```
When the transform is `image`, try to print the `node.role`

A partial exception is 
```
        Caused by:
        org.jruby.exceptions.RaiseException: (TypeError) cannot convert instance of class org.jruby.RubyString to boolean
            at org.jruby.gen.InterfaceImpl1017458621.isRole(org/jruby/gen/InterfaceImpl1017458621.gen:13)
            at org.ysb33r.asciidoctor.leanpub.LeanpubConverter.convertImage(LeanpubConverter.groovy:353)
            at org.ysb33r.asciidoctor.AbstractTextConverter.convert(AbstractTextConverter.groovy:102)
            at org.asciidoctor.converter.ConverterProxy.convert(org/asciidoctor/converter/ConverterProxy.java:90)
            at RUBY.convert(/Users/schalkc/.m2/repository/org/asciidoctor/asciidoctorj/1.6.0-SNAPSHOT/asciidoctorj-1.6.0-SNAPSHOT.jar!/gems/asciidoctor-1.5.2/lib/asciidoctor/abstract_block.rb:71)
            at RUBY.content(/Users/scha
```

BTW it does not fail for the normal converters (which are all in Ruby).

**Comment by robertpanzer** [comment]
It seems that `AbstractNode.isRole()` doesn't delegate to `AbstractNode::role?` but to the property reader `AbstractNode::role`.
I guess that if it returns `nil` it is converted to false but it can't do that if the role is set.

Not sure how to deal with that.
Copy the logic of `AbstractNode::role?`into `AbstractNode.isRole()`?
Or use Ruby reflection powers to explicitely invoke `role?` on the Ruby object instead of relying on JRubys Ruby/Java mapping magic.

**Comment by robertpanzer** [comment]
But it's definitely a different problem.

**Comment by mojavelinux** [comment]
> It seems that AbstractNode.isRole() doesn't delegate to AbstractNode::role? but to the property reader AbstractNode::role.

Wow, if that's the case than that's really broken in JRuby. Is there any way to tell JRuby not to do this? It really should look for a method first, then the property if there is no method.

**Comment by ysb33r** [comment]
> But it's definitely a different problem.

I've logged #323 to cover this case.

**Comment by robertpanzer** [comment]
Fixed with #317
